subsys/net/ip/connection.c

```diff
@@ -471,42 +478,44 @@ static inline void conn_send_icmp_error(struct net_pkt *pkt)
 	}
 }
 
-static bool conn_are_end_points_valid(struct net_pkt *pkt,
-				      union net_ip_header *ip_hdr,
-				      uint16_t src_port,
-				      uint16_t dst_port)
+#if defined(CONFIG_NET_IP)
+static bool conn_are_endpoints_valid(struct net_pkt *pkt, uint8_t family,
+				     union net_ip_header *ip_hdr,
+				     uint16_t src_port, uint16_t dst_port)
 {
-	bool my_src_addr = false;
-
-	/* For AF_PACKET family, we are not parsing headers. */
-	if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET) &&
-	    net_pkt_family(pkt) == AF_PACKET) {
+	bool is_my_src_addr;
+	bool is_same_src_and_dst_addr;
+
+	switch (family) {
+#if defined(CONFIG_NET_IPV4)
+	case AF_INET:
+		is_my_src_addr = net_ipv4_is_my_addr(
+			(struct in_addr *)ip_hdr->ipv4->src);
+		is_same_src_and_dst_addr = net_ipv4_addr_cmp_raw(
+			ip_hdr->ipv4->src, ip_hdr->ipv4->dst);
+		break;
+#endif
+#if defined(CONFIG_NET_IPV6)
+	case AF_INET6:
+		is_my_src_addr = net_ipv6_is_my_addr(
+			(struct in6_addr *)ip_hdr->ipv6->src);
+		is_same_src_and_dst_addr = net_ipv6_addr_cmp_raw(
+			ip_hdr->ipv6->src, ip_hdr->ipv6->dst);
+		break;
+#endif
+	default:
 		return true;
 	}
 
-	if (IS_ENABLED(CONFIG_NET_SOCKETS_CAN) &&
-	    net_pkt_family(pkt) == AF_CAN) {
-		return true;
-	}
+	bool is_same_src_and_dst_port = src_port == dst_port;
+	bool are_invalid_endpoints =
+		(is_same_src_and_dst_addr || is_my_src_addr) && is_same_src_and_dst_port;
 
-	if (IS_ENABLED(CONFIG_NET_IPV4) && net_pkt_family(pkt) == AF_INET) {
-		if (net_ipv4_addr_cmp_raw(ip_hdr->ipv4->src,
-					  ip_hdr->ipv4->dst) ||
-		    net_ipv4_is_my_addr((struct in_addr *)ip_hdr->ipv4->src)) {
-			my_src_addr = true;
-		}
-	} else if (IS_ENABLED(CONFIG_NET_IPV6) &&
-		   net_pkt_family(pkt) == AF_INET6) {
-		if (net_ipv6_addr_cmp_raw(ip_hdr->ipv6->src,
-					  ip_hdr->ipv6->dst) ||
-		    net_ipv6_is_my_addr((struct in6_addr *)ip_hdr->ipv6->src)) {
-			my_src_addr = true;
-		}
-	}
-
-	return !(my_src_addr && (src_port == dst_port));
+	return !are_invalid_endpoints;
 }
+#endif /* CONFIG_NET_IP */
 
+#if defined(CONFIG_NET_SOCKETS_PACKET)
```

Comment:
Similar case here. If we add the `IS_ENABLED(CONFIG_NET_SOCKETS_PACKET)` where this function `conn_raw_socket()` is called, then this ifdef check can be removed and this function gets compile tested most of the time.

Reply:
Fixed here as well as in two similar cases in the same file.